### DISCUSSION TITLE
Can CsvHelper store the CSV row number in the mapped object?

### DISCUSSION THREAD

**Discussion by reloaded** [JoshClose/CsvHelper#1689] [Q&A] [answered]
Hello!

I'm mapping my CSV records to a concrete object and I was wondering if there's a way to persist the row number in the concrete object?

I have a mapped object that looks like this:

```
public class CsvRowDto {
  public string FirstName {get; set;}
  public string LastName {get; set;}
}
```

And I'm trying to achieve something like this:

```
public class CsvRowDto {
  public int RowNumber {get; set;}
  public string FirstName {get; set;}
  public string LastName {get; set;}
}
```

The use case for this is to provide better error reporting. When my application detects a validation/business rule error I'd like to report the row number that had an error.

```
var records = csvReader.GetRecords<CsvRowDto>().ToList();

foreach(var person in records) {
  if(string.IsNullOrEmpty(person.FirstName)) {
    Console.WriteLine("First name is invalid: " + person.RowNumber);
  }
}
```

**Comment by JoshClose** [comment]
Text file row number or CSV row number? Meaning, a CSV field may have a `\r\n` in it.

**Comment by reloaded** [comment]
That's a quick response! :)

I think CSV row number?... I'm not certain on the differences between the two. The text file row number includes blank lines, right?

Let's say that you are a end user of my application and you upload a CSV. When my system detects an error it needs to tell you which row you should take a look at in the CSV file to find and fix the issue....

So... I think I need the CSV row number?

**Comment by JoshClose** [comment]
You'll want the file's line number then. A single CSV row may span many lines because a field could contain a '\r\n'.

```cs
void Main()
{
    using (var stream = new MemoryStream())
    using (var writer = new StreamWriter(stream))
    using (var reader = new StreamReader(stream))
    using (var csv = new CsvReader(reader))
    {
        writer.WriteLine("Id,Name");
        writer.WriteLine("1,one");
        writer.WriteLine("2,two");
        writer.Flush();
        stream.Position = 0;
        
        csv.Configuration.RegisterClassMap<TestMap>();
        csv.GetRecords<Test>().ToList().Dump();
    }
}

public class Test
{
    public int Id { get; set; }
    public string Name { get; set; }
    public int RowNumber { get; set; }
}

public sealed class TestMap : ClassMap<Test>
{
    public TestMap()
    {
        Map(m => m.Id);
        Map(m => m.Name);
        Map(m => m.RowNumber).ConvertUsing(row => row.Context.RawRow);
    }
}
```

**Comment by JoshClose** [comment]
Also, when an exception occurs, that information is available in the exception too.

**Comment by reloaded** [comment]
Wonderful! @JoshClose  That solved my question. Using `Map(m => m.RowNumber).ConvertUsing(row => row.Context.RawRow);` works like a charm.

**Comment by itscodeminer** [comment]
Hi, 
If the content of records span across multiples rows, the actual row number in the document is returned instead of record number, how do we get record number in csv in such a case
Thank you!

**Comment by JoshClose** [comment]
Use `Row` instead of `RawRow`.

**Comment by capinredbeard** [comment]
Is this still an option? It looks like the API has changed, and ConvertUsing is no longer available.  I have also tried to use a TypeConverter, but haven't found a way to get the row number.

EDIT: Just found the following, which appears to be what I need (untested):
Map(m => m.LineNumber).Convert(row => row.Context.Parser.RawRow);

**Comment by JoshClose** [comment]
Yes, `ConvertUsing` was renamed to `Convert`.

**Comment by jzabroski** [comment, marked as answer]
@JoshClose I think you should convert this to a Discussion and select the answer so that you can update the answer as you progress CsvHelper.

e.g.,

v23 (current):
```cs
Map(m => m.LineNumber).Convert(record => record.Row.Parser.Row);
```

v21:
```c#
Map(m => m.LineNumber).Convert(row => row.Context.Row);
```

v13 - v20:
```c#
Map(m => m.LineNumber).ConvertUsing(row => row.Context.Row);
```

v12:
```c#
Map(m => m.LineNumber).ConvertUsing(row => row.Context.RawRow);
```

**Reply by josedbaez** [reply]
This is the way in v23.

`Map(m => m.LineNumber).Convert(record => record.Row.Parser.Row);`

**Reply by bolorundurowb** [reply]
For some reason, this `Map(x => x.SerialNo).Name("S/No").Index(0).Convert(x => x.Row.Parser.Row + 1);` just leads to `0`'s in the output csv

**Reply by JoshClose** [reply]
You're specifying a name and index. If you're using `Convert` you should not specify those.

**Reply by IanKemp** [reply]
Suggestions:

* Update answer to have latest first, i.e. v21 then v13 - v20 then v12
* Update v21 to say "v21 - current"

**Reply by JoshClose** [reply]
Done.

**Reply by jzabroski** [reply]
Thank you.

**Reply by danielekiyoyo** [reply]
Thanks @JoshClose

**Reply by Venumadhavi-stack** [reply]
Hi,
How to fetch the row number from csv file if the row contains invalid columns .....after performing validation.
Can u give the sample code.

here what I have done is,
I read a csv file into a list of objects and using if condition performed field validations. if all field validations are success, wrote the values into one text file and if validations failed that column name should be written into the another text file along with the row value that has error.

plzz give me sample code for this problem as soon as possible.

Thanks in advance.
[POC 3(Algorithm).txt](https://github.com/JoshClose/CsvHelper/files/8951088/POC.3.Algorithm.txt)

**Comment by ajtruckle** [comment]
So it is not clear. Can I get the line number from the input document based on using attributes? I am not using a map. I appreciate error logging includes this info but sometime I show my own errors.

**Reply by Venumadhavi-stack** [reply]
Hi,
How to fetch the row number from csv file if the row contains invalid columns .....after performing validation.
Can u give the sample code.

here what I have done is,
I read a csv file into a list of objects and using if condition performed field validations. if all field validations are success, wrote the values into one text file and if validations failed that column name should be written into the another text file along with the row value that has error.

plzz give me sample code for this problem as soon as possible.

Thanks in advance.
[POC 3(Algorithm).txt](https://github.com/JoshClose/CsvHelper/files/8951072/POC.3.Algorithm.txt)

**Reply by JoshClose** [reply]
You could write your own attribute to do this.

```cs
void Main()
{
    var s = new StringBuilder();
    s.Append("Id,Name\r\n");
    s.Append("1,one\r\n");
    s.Append("2,two\r\n");
    var config = new CsvConfiguration(CultureInfo.InvariantCulture)
    {
    };
    using (var reader = new StringReader(s.ToString()))
    using (var csv = new CsvReader(reader, config))
    {
        csv.GetRecords<Foo>().ToList().Dump();
    }
}

private class Foo
{
    public int Id { get; set; }
    public string Name { get; set; }
    [LineNumber]
    public int LineNumber { get; set; }
}

private class LineNumber : Attribute, IMemberMapper, IParameterMapper
{
    public void ApplyTo(MemberMap memberMap)
    {
        memberMap.Data.ReadingConvertExpression = (ConvertFromStringArgs args) => args.Row.Parser.Row;
    }

    public void ApplyTo(ParameterMap parameterMap)
    {
    }
}
```
![image](https://user-images.githubusercontent.com/114971/175115889-7615ea80-56ba-40a2-90b0-e473989e2f34.png)

**Reply by Venumadhavi-stack** [reply]
I wrote  a code for reading csv file into list of objects. And by using foreach I validated each column and if all records in a row satisfied with validations it should write the whole in one text file . If one record got failed with validation it should write the entry in another file along with the row number. 
Now , plz tell me how can I retrieve the row number.using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.IO;
using System.Linq;

namespace Example
{
    class Program
    {
        

        static void Main(string[] args)
        {
            List<Header> header = File.ReadAllLines("C:\\Users\\GVM\\source\\repos\\Input2.csv")
                                          .Skip(1)
                                          .Select(v => Header.FromCsv(v))

                                          .ToList();
            foreach(var i in header)
            {
                if((!String.IsNullOrEmpty(i.sourceDate)) && i.sourceDate.Length<=16 && (!String.IsNullOrEmpty(i.sourceLine)) && i.sourceLine.Length<=9 && (!String.IsNullOrEmpty(i.journalID)) && i.journalID.Length<=10 && (!String.IsNullOrEmpty(i.accountingDate)) && i.accountingDate.Length<=10 && (!String.IsNullOrEmpty(i.businessUnit)) && i.businessUnit.Length<=5 && (!String.IsNullOrEmpty(i.ledgerGroup)) && i.ledgerGroup.Length<=10 && (!String.IsNullOrEmpty(i.ledger)) && i.ledger.Length<=10 && (!String.IsNullOrEmpty(i.account)) && i.account.Length<=10 && (!String.IsNullOrEmpty(i.alternateAccount)) && i.alternateAccount.Length<=10&&(!String.IsNullOrEmpty(i.operatingUnit))&&i.operatingUnit.Length<=8 && (!String.IsNullOrEmpty(i.costCentre)) && i.costCentre.Length<=10 && (!String.IsNullOrEmpty(i.product)) && i.product.Length <= 6 && (!String.IsNullOrEmpty(i.year)) && i.year.Length <= 4 && (!String.IsNullOrEmpty(i.mcc)) && i.mcc.Length<=7 && (!String.IsNullOrEmpty(i.geoCode)) && i.geoCode.Length<=5 && (!String.IsNullOrEmpty(i.function)) && i.function.Length<=2 && (!String.IsNullOrEmpty(i.projectID)) && i.projectID.Length<=15 && (!String.IsNullOrEmpty(i.affiliate)) && i.affiliate.Length<=5 && (!String.IsNullOrEmpty(i.openitemKey)) && i.openitemKey.Length<=10 && (!String.IsNullOrEmpty(i.foreignCurrency)) && i.foreignCurrency.Length<=3 && (!String.IsNullOrEmpty(i.currencyCode)) && i.currencyCode.Length<=3 && (!String.IsNullOrEmpty(i.foreignAmount)) && i.foreignAmount.Length<=27 && (!String.IsNullOrEmpty(i.monetaryAmount)) && i.monetaryAmount.Length<=27 && (!String.IsNullOrEmpty(i.docType)) && i.docType.Length<=8 && (!String.IsNullOrEmpty(i.docSeqNbr)) && i.docSeqNbr.Length<=12 && (!String.IsNullOrEmpty(i.docSeqDate)) && i.docSeqDate.Length<=10 && (!String.IsNullOrEmpty(i.distribStatus)) && i.distribStatus.Length==1 && (!String.IsNullOrEmpty(i.epmExtracted)) && i.epmExtracted.Length==1 && (!String.IsNullOrEmpty(i.quantumExtra1)) && i.quantumExtra1.Length<=9 && (!String.IsNullOrEmpty(i.quantumExtra2)) && i.quantumExtra2.Length<=40 && (!String.IsNullOrEmpty(i.quantumExtra3)) && i.quantumExtra3.Length<=10 && (!String.IsNullOrEmpty(i.quantumExtra4)) && i.quantumExtra4.Length<=20 && (!String.IsNullOrEmpty(i.quantumExtra5)) && i.quantumExtra5.Length<=10 && (!String.IsNullOrEmpty(i.analyse01)) && i.analyse01.Length<=35 && (!String.IsNullOrEmpty(i.analyse02)) && i.analyse02.Length<=35)
                {
                using(StreamWriter writer=new StreamWriter(@"C:\Users\GVM\source\repos\Actual.txt", true))
                    {
                        writer.WriteLine("{0,-16}{1,-9}{2,-10}{3,-10}{4,-5}{5,-10}{6,-10}{7,-10}{8,-10}{9,-8}{10,-10}{11,-6}{12,-4}{13,-7}{14,-5}{15,-2}{16,-15}{17,-5}{18,-10}{19,-3}{20,-3}{21,-27}{22,-27}{23,-8}{24,-12}{25,-10}{26,-1}{27,-1}{28,-9}{29,-40}{30,-10}{31,-20}{32,-10}{33,-35}{34,-35}", i.sourceDate, i.sourceLine, i.journalID, i.accountingDate, i.businessUnit, i.ledgerGroup, i.ledger, i.account, i.alternateAccount, i.operatingUnit, i.costCentre, i.product, i.year, i.mcc, i.geoCode, i.function, i.projectID, i.affiliate, i.openitemKey, i.foreignCurrency, i.currencyCode, i.foreignAmount, i.monetaryAmount, i.docType, i.docSeqNbr, i.docSeqDate, i.distribStatus, i.epmExtracted, i.quantumExtra1, i.quantumExtra2, i.quantumExtra3, i.quantumExtra4, i.quantumExtra5, i.analyse01, i.analyse02);
                    }
                }
                else
                {
                    

                        if (String.IsNullOrEmpty(i.sourceDate))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("SourceDate has  null value");
                        }
                        if (i.sourceDate.Length > 16)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("SourceDate has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.sourceLine))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("SourceLine has  null value");
                        }
                        if (i.sourceLine.Length > 9)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("SourceLine has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.journalID))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("journalID has  null value");
                        }
                        if (i.journalID.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("JournalID has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.accountingDate))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("accountingDate has  null value");
                        }
                        if (i.accountingDate.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("accountingDate has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.businessUnit))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("businessUnit has  null value");
                        }
                        if (i.businessUnit.Length > 5)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine(" businessUnit has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.ledgerGroup))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("ledgerGroup has  null value");
                        }
                        if (i.ledgerGroup.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("ledgerGroup has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.ledger))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("ledger has  null value");
                        }
                        if (i.ledger.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("ledger has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.account))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("account has  null value");
                        }
                        if (i.account.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("account has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.alternateAccount))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("alternateAccount has  null value");
                        }
                        if (i.alternateAccount.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("alternateAccount has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.operatingUnit))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("operatingunit has  null value");
                        }
                        if (i.operatingUnit.Length > 8)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("operatingUnit has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.costCentre))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("costCentre has  null value");
                        }
                        if (i.costCentre.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("costCentre has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.product))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("product has  null value");
                        }
                        if (i.product.Length > 6)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("product has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.year))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("year has  null value");
                        }
                        if (i.year.Length > 4)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("year has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.mcc))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("mcc has  null value");
                        }
                        if (i.mcc.Length > 7)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("mcc has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.geoCode))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("geoCode has  null value");
                        }
                        if (i.geoCode.Length > 5)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("geoCode has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.function))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("function has  null value");
                        }
                        if (i.function.Length > 2)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("function has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.projectID))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("projectID has  null value");
                        }
                        if (i.projectID.Length > 15)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("projectID has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.affiliate))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("affiliate has  null value");
                        }
                        if (i.affiliate.Length > 5)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("affiliate has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.openitemKey))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("openitemKey has  null value");
                        }
                        if (i.openitemKey.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("openitemKey has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.foreignCurrency))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("foreignCurrency has  null value");
                        }
                        if (i.foreignCurrency.Length > 3)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("foreignCurrency has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.currencyCode))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("currencyCode has  null value");
                        }
                        if (i.currencyCode.Length > 3)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("currencyCode has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.foreignAmount))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("foreignAmount has  null value");
                        }
                        if (i.foreignAmount.Length > 27)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("foreignAmount has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.monetaryAmount))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("monetaryAmount has  null value");
                        }
                        if (i.monetaryAmount.Length > 27)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("monetaryAmount has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.docType))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("docType has  null value");
                        }
                        if (i.docType.Length > 8)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("docType has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.docSeqNbr))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("docSeqNbr has  null value");
                        }
                        if (i.docSeqNbr.Length > 12)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("docSeqNbr has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.docSeqDate))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("docSeqDate has  null value");
                        }
                        if (i.docSeqDate.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("docSeqDate has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.distribStatus))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("distribStatus has  null value");
                        }
                        if (i.distribStatus.Length > 1)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("distribStatus has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.epmExtracted))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("epmExtracted has  null value");
                        }
                        if (i.epmExtracted.Length > 1)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("epmExtracted has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.quantumExtra1))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("quantumExtra1 has  null value");
                        }
                        if (i.quantumExtra1.Length > 9)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("quantumExtra1 has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.quantumExtra2))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("quantumExtra2 has  null value");
                        }
                        if (i.quantumExtra2.Length > 40)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("quantumExtra2 has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.quantumExtra3))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("quantumExtra3 has  null value");
                        }
                        if (i.quantumExtra3.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("quantumExtra3 has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.quantumExtra4))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("quantumExtra4 has  null value");
                        }
                        if (i.quantumExtra4.Length > 20)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("quantumExtra4 has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.quantumExtra5))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("quantumExtra5 has  null value");
                        }
                        if (i.quantumExtra5.Length > 10)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("quantumExtra5 has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.analyse01))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("analyse01 has  null value");
                        }
                        if (i.analyse01.Length > 35)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("analyse01 has invalid length");
                        }
                        if (String.IsNullOrEmpty(i.analyse02))
                        {

                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("analyse02 has  null value");
                        }
                        if (i.analyse02.Length > 35)

                        {
                            using (StreamWriter error = new StreamWriter(@"C:\Users\GVM\source\repos\ErrorDetails.txt", true))
                                error.WriteLine("analyse02 has invalid length");
                        }

                    
                }
            }
        }
        public class Header
        {
            public string sourceDate;
            public string sourceLine;
            public string journalID;
            public string accountingDate;
            public string businessUnit;
            public string ledgerGroup;
            public string ledger;
            public string account;
            public string alternateAccount;
            public string operatingUnit;
            public string costCentre;
            public string product;
            public string year;
            public string mcc;
            public string geoCode;
            public string function;
            public string projectID;
            public string affiliate;
            public string openitemKey;
            public string foreignCurrency;
            public string currencyCode;
            public string foreignAmount;
            public string monetaryAmount;
            public string docType;
            public string docSeqNbr;
            public string docSeqDate;
            public string distribStatus;
            public string epmExtracted;
            public string quantumExtra1;
            public string quantumExtra2;
            public string quantumExtra3;
            public string quantumExtra4;
            public string quantumExtra5;
            public string analyse01;
            public string analyse02;
            

            public static Header FromCsv(string csvLine)
            {
                string[] values = csvLine.Split(',');
                Header header = new Header();
                header.sourceDate = Convert.ToString(values[0]);
                //header.sourceDate.ToShortDateString(values[0]);
                
                header.sourceLine = Convert.ToString(values[1]);
                header.journalID = Convert.ToString(values[2]);
                header.accountingDate = Convert.ToString(values[3]);
                header.businessUnit = Convert.ToString(values[4]);
                header.ledgerGroup = Convert.ToString(values[5]);
                header.ledger = Convert.ToString(values[6]);
                header.account = Convert.ToString(values[7]);
                header.alternateAccount = Convert.ToString(values[8]);
                header.operatingUnit = Convert.ToString(values[9]);
                header.costCentre = Convert.ToString(values[10]);
                header.product = Convert.ToString(values[11]);
                header.year = Convert.ToString(values[12]);
                header.mcc = Convert.ToString(values[13]);
                header.geoCode = Convert.ToString(values[14]);
                header.function = Convert.ToString(values[15]);
                header.projectID = Convert.ToString(values[16]);
                header.affiliate = Convert.ToString(values[17]);
                header.openitemKey = Convert.ToString(values[18]);
                header.foreignCurrency = Convert.ToString(values[19]);
                header.currencyCode = Convert.ToString(values[20]);
                header.foreignAmount = Convert.ToString(values[21]);
                header.monetaryAmount = Convert.ToString(values[22]);
                header.docType = Convert.ToString(values[23]);
                header.docSeqNbr = Convert.ToString(values[24]);
                header.docSeqDate = Convert.ToString(values[25]);
                header.distribStatus = Convert.ToString(values[26]);
                header.epmExtracted = Convert.ToString(values[27]);
                header.quantumExtra1 = Convert.ToString(values[28]);
                header.quantumExtra2 = Convert.ToString(values[29]);
                header.quantumExtra3 = Convert.ToString(values[30]);
                header.quantumExtra4 = Convert.ToString(values[31]);
                header.quantumExtra5 = Convert.ToString(values[32]);
                header.analyse01 = Convert.ToString(values[33]);
                header.analyse02 = Convert.ToString(values[34]);
                return header;
            
            

               


            }



        }
    }
}
This is the code I wrote....Plz ans me ASAP.
Thanks in advance

**Reply by JoshClose** [reply]
@Venumadhavi-stack This isn't using CsvHelper.

**Reply by jsnape** [reply]
Should this work with records? I have implemented the above ` this.Map(m => m.RowNumber).Convert(static (args) => args.Row.Parser.RawRow);` but when it comes to creating records I end up with `System.ArgumentException : Incorrect number of arguments for constructor`.

```CSharp
public record AddressCsvRecord(
    string? Uprn,
    string? ClassificationCode,
    string? SingleLineAddress,
    string? Organisation,
    string? SubBuilding,
    string? BuildingName,
    string? BuildingNumber,
    string? StreetName,
    string? Locality,
    string? TownName,
    string? Postcode,
    int RowNumber = 0,
    ValidationResult? Validation = null);

// Class map constructor:

    public AddressCsvRecordMap()
    {
        this.AutoMap(CultureInfo.InvariantCulture);

        this.Map(m => m.RowNumber)
            .Convert(static (args) => args.Row.Parser.RawRow);
    }
```

With the following exception:
```
    CsvHelper.ReaderException : An unexpected error occurred.
IReader state:
   ColumnCount: 27
   CurrentIndex: -1
   HeaderRecord:
["UPRN","PARENT_UPRN","UDPRN","USRN","TOID","CLASSIFICATION_CODE","EASTING","NORTHING","LATITUDE","LONGITUDE","RPC","LAST_UPDATE_DATE","SINGLE_LINE_ADDRESS","PO_BOX","ORGANISATION","SUB_BUILDING","BUILDING_NAME","BUILDING_NUMBER","STREET_NAME","LOCALITY","TOWN_NAME","POST_TOWN","ISLAND","POSTCODE","DELIVERY_POINT_SUFFIX","GSS_CODE","CHANGE_CODE"]
IParser state:
   ByteCount: 0
   CharCount: 514
   Row: 2
   RawRow: 2
   Count: 27
   RawRecord:
10013036001,,8746390,14200018,"osgb1000012340791","RD",293750,92669,50.7238873,-3.5066416,1,2020-01-06,"28 ALPHA STREET, EXETER, EX1 2SP","","","","","28","ALPHA STREET","","EXETER","","","EX1 2SP","2X","E07000041","I"


---- System.ArgumentException : Incorrect number of arguments for constructor
```

I have tried defaults,  [mapping parameters](https://stackoverflow.com/questions/68143146/csvhelper-with-non-default-constructor), [removing a referencemap](https://github.com/JoshClose/CsvHelper/issues/1271)

Will I have to convert records to classes or is there another option?
